EIPS/eip-5727.md

```diff
@@ -0,0 +1,596 @@
+     * @dev MUST revert if the `index` exceed the total number of slots.
+     * @param index The index of the slot queried for
+     * @return The slot is queried for
+     */
+    function slotByIndex(uint256 index) external view returns (uint256);
+
+    /**
+     * @notice Get the number of tokens in a slot.
+     * @dev MUST revert if the slot does not exist.
+     * @param slot The slot whose number of tokens is queried for
+     * @return The number of tokens in the `slot`
+     */
+    function tokenSupplyInSlot(uint256 slot) external view returns (uint256);
+
+    /**
+     * @notice Get the tokenId with `index` of the `slot`.
+     * @dev MUST revert if the `index` exceed the number of tokens in the `slot`.
+     * @param slot The slot whose token is queried for.
+     * @param index The index of the token queried for
+     * @return The token is queried for
+     */
+    function tokenInSlotByIndex(uint256 slot, uint256 index)
+        external
+        view
+        returns (uint256);
+}
+```
+
+## Rationale
+### Token storage model
+We adopt semi-fungible token storage models designed to support both fungible and non-fungible tokens, inspired by [EIP-3525](./eip-3525.md). We found that such a model is better suited to the representation of SBT than the model used in [EIP-1155](./eip-1155.md).
+
+Firstly, each slot can be used to represent different categories of SBTs. For instance, a DAO can have membership SBTs, role badges, scores, etc. in one SBT collection.
+
+Secondly, unlike [EIP-1155](./eip-1155.md), in which each unit of fungible tokens is exactly the same, our interface can help differentiate between similar tokens. This is justified by that credential scores obtained from different entities differ not only in value but also in their effects, validity periods, origins, etc. However, they still share the same slot as they all contribute to a person's credibility.
+
+### Recovery mechanism
+To prevent the loss of SBTs, we propose a recovery mechanism that allows users to recover their tokens by providing a signature signed by their soul address. This mechanism is inspired by [EIP-1271](./eip-1271.md).
+
+Since SBTs are bound to an address and are meant to represent the identity of the address, which cannot be split into fractions. Therefore, each recovery should be considered as a transfer of all the tokens of the owner. This is why we use the `recover` function instead of `transferFrom` or `safeTransferFrom`.
+
+### Token visibility control
+Our interface allows users to control the visibility of their tokens (shadowing and revealing). This is useful when a user wants to hide some of their tokens from the public, for example, when they want to keep their membership secret. Generally, the issuer and the owner of the token have access to the token by default and can control the visibility of the token. After the token is shadowed, information about the token (e.g. token URI, owner of the token) cannot be queried by the public.
```

Comment:
All the information remains visible on the chain, right? So how do you prevent queries?

Reply:
Yes, one can still get the information by looking at the tx data. Here, preventing query means preventing query through the interface of the contract. For example, if someone issues you a weird random SBT, you can hide it so that it will not be listed by the frontend.

Reply:
Does that mean the holder of SBT will need to check what SBT is issued to him/her, and pay gas to hide some weird SBT?
That seems unpractical, as I receive a lot of garbage NFT these days and I will do nothing to them.

Reply:
Yes, I ignore those NFTs too. Although this may be a niche use case, I think being able to control the visibility might be interesting and convenient.